Doc/library/typing.rst

```diff
@@ -933,6 +933,84 @@ These can be used as types in annotations using ``[]``, each having a unique syn
+   type guard function.  ``TypeGuard`` only accepts a single type argument.
+   At runtime, functions marked this way should return a boolean.
+
+   ``TypeGuard`` aims to benefit *type narrowing* - a technique used by static
+   type checkers to determine a more precise type of an expression within a
+   program's code flow.  Usually type narrowing is done by analyzing
+   conditional code flow and applying the narrowing to a block of code.  The
+   conditional expression here is sometimes referred to as a "type guard"::
+
+      def func(val: Optional[Union[str, float]]):
+          # Non-"None" type guard
+          if val is not None:
+              # Type of val is narrowed to ``Union[str, float]``
+              # "isinstance" type guard
+              if isinstance(val, str):
+                  # Type of val is narrowed to ``str``
+                  ...
+              else:
+                  # Else, type of val is narrowed to ``float``.
+                  ...
+          else:
+              # Type of val remains Optional[Union[str, float]]
+              ...
+
+   Sometimes, a type guard uses a user-defined checking function instead of
+   ``isinstance`` or ``is None`` checks.  These user-defined type guard
+   functions require ``TypeGuard`` to narrow their input types as the static
+   type checker usually does not have enough information to statically infer
+   them.
+
+   Using  ``-> TypeGuard`` tells the static type checker that for a given
+   function:
+
+   1. The return value is a boolean.
+   2. If the return value is ``True``, the type of its argument
+      is the type inside ``TypeGuard``.
+
+      For example::
+
+         def is_str_list(val: List[object]) -> TypeGuard[List[str]]:
+             '''Determines whether all objects in the list are strings'''
+             return all(isinstance(x, str) for x in val)
+
+         def func1(val: List[object]):
+             if is_str_list(val):
+                 # Type of ``val`` is narrowed to List[str]
+                 print(" ".join(val))
+             else:
+                 # Type of ``val`` remains as List[object]
+                 print("Not a list of strings!")
+
+   If ``is_str_list`` is a class or instance method, then the type in
+   ``TypeGuard`` maps to the type of the second parameter after ``cls`` or
+   ``self``.
+
+   In short, the form ``def foo(arg: TypeA) -> TypeGuard[TypeB]: ...``,
+   means that if ``foo(arg)`` returns ``True``, then ``arg`` narrows from
+   ``TypeA`` to ``TypeB``.
+
+   .. note::
+
+      Strict type narrowing is not enforced - ``TypeB`` need not be a narrower
+      form of ``TypeA`` (it can even be a wider form) and this may lead to
+      type-unsafe results.  The main reason is to allow for things like
+      narrowing ``List[object]`` to ``List[str]`` which would fail under strict
+      narrowing as ``List`` is invariant.  The responsibility of
+      writing type-safe type guards is left to the user.
```

Comment:
```suggestion
      ``TypeB`` need not be a narrower
      form of ``TypeA`` -- it can even be a wider form.
      The main reason is to allow for things like
      narrowing ``List[object]`` to ``List[str]`` even though the latter
      is not a subtype of the former, since ``List`` is invariant.  The responsibility of
      writing type-safe type guards is left to the user.
```